### PR TITLE
Additional constraints to prevent violation of state of charge limits in non-representative periods

### PR DESCRIPTION
**Pull request by federicoparolin**
## Description

- Introduced additional constraints to bound the state of charge of LDS systems within the limits (zero and installed energy capacity) throughout the entire time horizon (as described in https://doi.org/10.48550/arXiv.2409.19079).
- Saving of storage content evolution throughout the entire time horizon (StorageEvol.csv and HSC_StorageEvol.csv)

Storage content evolution over full time horizon with the original (a) and updated (b) formulations:
immagine

## What type of PR is this? (check all applicable)

- [x] Feature
- [x] Bug Fix
- [ ] Documentation Update
- [ ] Code Refactor
- [ ] Performance Improvements

## How this can be tested

Tested on "3_three_zones_w_co2_capture" enabling LDS on storage resources.

## Post-approval checklist for GenX core developers
After the PR is approved

- [x] Check that the latest changes on the target branch are incorporated, either via merge or rebase
- [x] Remember to squash and merge if incorporating into develop

### CODE REVIEW

**Review by gmantegna — CHANGES_REQUESTED** [review]
Looks good overall, just two requests: 1. Can you please modify the relevant doc pages to reflect this change? 2. Do you have an idea of how much this increases runtime? Especially for a large case?

**Review by gmantegna — APPROVED** [review]
Looks good, thank you

File: src/write_outputs/long_duration_storage/write_opwrap_lds_stor_init.jl [hunk 1]
```diff
@@ -32,4 +32,52 @@ function write_opwrap_lds_stor_init(path::AbstractString,
...
+    t_interior = 2:hours_per_subperiod
+    T = hours_per_subperiod*NPeriods
+    SOC_t = zeros(G, T)
+    stor_long_duration = inputs["STOR_LONG_DURATION"]
+    stor_hydro_long_duration = inputs["STOR_HYDRO_LONG_DURATION"]
+    period_map = inputs["Period_Map"].Rep_Period_Index
+    pP_max = inputs["pP_Max"]
+    e_total_cap = value.(EP[:eTotalCap])
+    v_charge = value.(EP[:vCHARGE])*1000
+    v_P = value.(EP[:vP])*1000
```
Comment (lbonaldo) [inline src/write_outputs/long_duration_storage/write_opwrap_lds_stor_init.jl:47]: Is this multiplication intentional? Is it related to the scaling factor? If so, I think we should do something like:
```suggestion
    v_charge = value.(EP[:vCHARGE])
    v_P = value.(EP[:vP])
    if setup["ParameterScale"] == 1
        v_charge *= ModelScalingFactor
        v_P *= ModelScalingFactor
    end
```

File: src/write_outputs/long_duration_storage/write_opwrap_lds_stor_init.jl [hunk 2]
```diff
@@ -32,4 +32,52 @@ function write_opwrap_lds_stor_init(path::AbstractString,
...
+            SOC_t[stor_hydro_long_duration, t_r] = socw[stor_hydro_long_duration, r] .- 1 ./ efficiency_down.(gen[stor_long_duration]) .* v_P[stor_hydro_long_duration, t_start_w] .- v_spill[stor_hydro_long_duration, t_start_w] .+ pP_max[stor_hydro_long_duration, t_start_w] .* e_total_cap[stor_hydro_long_duration]
+
+            for t_int in t_interior
+                t = hours_per_subperiod * (w - 1) + t_int
+                SOC_t[stor_hydro_long_duration, t_r + t_int - 1] = SOC_t[stor_hydro_long_duration, t_r + t_int - 2] .- 1 ./ efficiency_down.(gen[stor_long_duration]) .* v_P[stor_hydro_long_duration, t] .- v_spill[stor_hydro_long_duration, t] .+ pP_max[stor_hydro_long_duration, t] .* e_total_cap[stor_hydro_long_duration]
+            end
+        end
+    end
+    # println(SOC_t[10,:])
+    # error("STOP")
```
Comment (lbonaldo) [inline src/write_outputs/long_duration_storage/write_opwrap_lds_stor_init.jl:76]: No longer needed...
```suggestion
```

File: src/write_outputs/long_duration_storage/write_opwrap_lds_stor_init.jl [hunk 3]
```diff
@@ -32,4 +32,52 @@ function write_opwrap_lds_stor_init(path::AbstractString,
...
+    df_SOC_t = DataFrame(Resource = inputs["RESOURCE_NAMES"], Zone = zones)
+    df_SOC_t = hcat(df_SOC_t, DataFrame(SOC_t, :auto))
+    auxNew_Names = [Symbol("Resource"); Symbol("Zone"); [Symbol("n$t") for t in 1:T]]
```
Comment (lbonaldo) [inline src/write_outputs/long_duration_storage/write_opwrap_lds_stor_init.jl:79]: In the codebase, `T` typically refers to the number of time steps read from the `demand` file (which is stored in `inputs["T"]`). If I’m correct, the T at this line represents the length of the sub-periods multiplied by their count (and can therefore be different from `inputs["T"]`). Could you please use a different name to improve code clarity?

**Review by lbonaldo — CHANGES_REQUESTED** [review]
Hey @federicoparolin, thanks so much for this PR! It looks great to me as well. I’ve added a few additional comments on the `write_opwrap_lds_stor_init` function, could you please take a look when you have a moment?

**Review by lbonaldo — APPROVED** [review]

**Pull request merged**